### PR TITLE
feat(ingest): Add metabase database id to platform instance mapping

### PR DESCRIPTION
**Pull request by k-popov**
## Checklist

- [X] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [X] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

Since DataHub supports [platform instances](https://datahubproject.io/docs/platform-instances/) there may be several distinct instances with the same platform in the same environment. When ingesting charts from Metabase it should be possible to find the corresponding datasets (tables) in DataHub.

For example, the dataset URN searched during ingestion is:
- *without* the patch: `urn:li:dataPlatform:clickhouse,event.new_app,PROD`
- *with* the patch: `urn:li:dataPlatform:clickhouse,clichouse_analytics_prod.event.new_app,PROD`
with only the latter URN being valid for dataset ingested with `platform_instance` specified.

The change still leaves `name_to_instance_map` optional and if `platform_instance` is not used, the mapping may be skipped.

Also the change adds a case for Metabase user not being found which is normal for deactivated users. I believe this case should not be considered as failure as people in companies quit or change their roles while charts and dashboards they created are still used by the company. I believe this behavior has already been mentioned in #5294. Unfortunately Metabase API does not support GET'ting individual user details for deactivated users, only has options to [list them](https://github.com/metabase/metabase/blob/master/docs/api/user.md#get-apiuserid). Thus checking for 404 looks like a reasonable and yet simple workaround.

### CODE REVIEW

**Comment by k-popov** [comment]
Another option for mapping would be using `id` of database in Metabase instead of its name. So if this is decided to be a better option, I will update the pull request.

**Comment by hsheth2** [comment]
@k-popov thanks for the contribution. @mayurinehate could you take a look?

**Comment by mayurinehate** [comment]
Hey @k-popov can you not use `platform_instance_map` config already available in metabase connector config ? for example - 
```

platform_instance_map:
    clickhouse: clichouse_analytics_prod
```

Or do you have multiple platform instances of clickhouse accessed via same metabase instance ?

**Comment by k-popov** [comment]
@mayurinehate thanks for feedback!

Yes, in my case there are multiple platform instances of clickhouse all accessed by the same metabase instance.
In my understanding that's quite a common case in larger companies: different product apps are using separate database instances but there is a single point of access to them - metabase.

I first thought of "re-using" the `platform_instance_map` config parameter but it would then be a breaking change which I'd prefer to avoid. That's why I introduced a new config option. Still I'm in doubt about the key for the mapping: database `id` is forced to be unique by Metabase while `name` can possibly have duplicates. On the other hand mapping with names is more "user-friendly" than mapping with numeric id. What option would you recommend?

**Comment by mayurinehate** [comment]
> @mayurinehate thanks for feedback!
> 
> Yes, in my case there are multiple platform instances of clickhouse all accessed by the same metabase instance. In my understanding that's quite a common case in larger companies: different product apps are using separate database instances but there is a single point of access to them - metabase.
> 
> I first thought of "re-using" the `platform_instance_map` config parameter but it would then be a breaking change which I'd prefer to avoid. That's why I introduced a new config option. Still I'm in doubt about the key for the mapping: database `id` is forced to be unique by Metabase while `name` can possibly have duplicates. On the other hand mapping with names is more "user-friendly" than mapping with numeric id. What option would you recommend?

Got it. I am inclined towards using id than name as that would solve the problem permanently and handle corner cases as well. About user friendliness, can you confirm if this is still true - https://github.com/datahub-project/datahub/pull/5647#issuecomment-1358880948 ? If yes, I'm comfortable with using database id. - and the config becomes `database_id_to_platform_instance_map`

**Comment by k-popov** [comment]
@mayurinehate as discussed, I've replaced key for mapping from `name` to `id`.
The key has been set to be string, not integer like it is in Metabase API as integer keys seem to not be supported by DataHub config: when I tried to make the key integer, `datahub ingest` has thrown `AttributeError: 'int' object has no attribute 'endswith'`. That's why I replaced numeric config keys with string.

I did not squash commits yet but if that's required for contribution, please let me know.

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py
Comment (k-popov) [reply metadata-ingestion/src/datahub/ingestion/source/metabase.py]: I have replaced the `hasattr` check with `http_error.response is not None` as `http_error.response` is always set to `None` in exception constructor: https://github.com/psf/requests/blob/cdbc2e271529f467b278b2760f12ee0b5d6930d3/requests/exceptions.py#L19

**Comment by mayurinehate** [comment]
> @mayurinehate as discussed, I've replaced key for mapping from `name` to `id`. The key has been set to be string, not integer like it is in Metabase API as integer keys seem to not be supported by DataHub config: when I tried to make the key integer, `datahub ingest` has thrown `AttributeError: 'int' object has no attribute 'endswith'`. That's why I replaced numeric config keys with string.

I am not familiar with such limitation. This should be supported. Do you happen to have stack trace of the error, i.e. which line in code gave the error ? I am fine with keeping these as strings. I believe, it is not mandatory to enclose the id key within quotes in recipe yml.

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py
Comment (k-popov) [reply metadata-ingestion/src/datahub/ingestion/source/metabase.py]: Yes, `database_id_to_instance_map` looks more self-explanatory. Updated the variable name.

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py
Comment (k-popov) [reply metadata-ingestion/src/datahub/ingestion/source/metabase.py]: Done here and on two lines below in similar `.get()`

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py
Comment (k-popov) [reply metadata-ingestion/src/datahub/ingestion/source/metabase.py]: That was my original idea but after your comment I realized it was not going to work with my previous implementation. Thank you for noticing this! I've updated the code to implement this fallback.

File: metadata-ingestion/docs/sources/metabase/metabase.md
Comment (k-popov) [reply metadata-ingestion/docs/sources/metabase/metabase.md]: Now it is used like you described. This document has been updated too.

**Comment by k-popov** [comment]
> 

Here is the stacktrace I see when using `int` as config key:
```
[2023-07-13 21:28:17,685] WARNING  {datahub.ingestion.run.pipeline:296} - Failed to configure reporter: datahub
Traceback (most recent call last):
  File "/home/kpopov/github/datahub/venv/lib/python3.10/site-packages/datahub/ingestion/run/pipeline.py", line 282, in _configure_reporting
    reporter_class.create(
  File "/home/kpopov/github/datahub/venv/lib/python3.10/site-packages/datahub/ingestion/reporting/datahub_ingestion_run_summary_provider.py", line 111, in create
    return cls(sink, reporter_config.report_recipe, ctx)
  File "/home/kpopov/github/datahub/venv/lib/python3.10/site-packages/datahub/ingestion/reporting/datahub_ingestion_run_summary_provider.py", line 139, in __init__
    recipe=self._get_recipe_to_report(ctx),
  File "/home/kpopov/github/datahub/venv/lib/python3.10/site-packages/datahub/ingestion/reporting/datahub_ingestion_run_summary_provider.py", line 156, in _get_recipe_to_report
    return json.dumps(redact_raw_config(ctx.pipeline_config._raw_dict))
  File "/home/kpopov/github/datahub/venv/lib/python3.10/site-packages/datahub/configuration/common.py", line 60, in redact_raw_config
    return {
  File "/home/kpopov/github/datahub/venv/lib/python3.10/site-packages/datahub/configuration/common.py", line 61, in <dictcomp>
    k: _redact_value(v) if _should_redact_key(k) else redact_raw_config(v)
  File "/home/kpopov/github/datahub/venv/lib/python3.10/site-packages/datahub/configuration/common.py", line 60, in redact_raw_config
    return {
  File "/home/kpopov/github/datahub/venv/lib/python3.10/site-packages/datahub/configuration/common.py", line 61, in <dictcomp>
    k: _redact_value(v) if _should_redact_key(k) else redact_raw_config(v)
  File "/home/kpopov/github/datahub/venv/lib/python3.10/site-packages/datahub/configuration/common.py", line 60, in redact_raw_config
    return {
  File "/home/kpopov/github/datahub/venv/lib/python3.10/site-packages/datahub/configuration/common.py", line 61, in <dictcomp>
    k: _redact_value(v) if _should_redact_key(k) else redact_raw_config(v)
  File "/home/kpopov/github/datahub/venv/lib/python3.10/site-packages/datahub/configuration/common.py", line 60, in redact_raw_config
    return {
  File "/home/kpopov/github/datahub/venv/lib/python3.10/site-packages/datahub/configuration/common.py", line 61, in <dictcomp>
    k: _redact_value(v) if _should_redact_key(k) else redact_raw_config(v)
  File "/home/kpopov/github/datahub/venv/lib/python3.10/site-packages/datahub/configuration/common.py", line 35, in _should_redact_key
    return key in REDACT_KEYS or any(key.endswith(suffix) for suffix in REDACT_SUFFIXES)
  File "/home/kpopov/github/datahub/venv/lib/python3.10/site-packages/datahub/configuration/common.py", line 35, in <genexpr>
    return key in REDACT_KEYS or any(key.endswith(suffix) for suffix in REDACT_SUFFIXES)
AttributeError: 'int' object has no attribute 'endswith'
```
Actually ingestion continues after this stacktrace but having any exception seems bad.

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/metabase.py:595]: This if needs to be written as below for this to behave correctly for case platform_instance was chosen from `datasource_id_in_metabase` .

```
if self.config.platform_instance_map and platform_instance is None:
     platform_instance = self.config.platform_instance_map.get(platform)
```

Can you please refractor the entire get platform instance logic in separate method and add unit tests for that method ?

```
def get_platform_instance_for_datasource(datasource_id: str, platform: Optional[str])->Optional[str]:
    ...
```

**Review by mayurinehate — CHANGES_REQUESTED** [review]

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py
Comment (k-popov) [reply metadata-ingestion/src/datahub/ingestion/source/metabase.py]: I have moved platform_instance detection into method.
Attempted to add a unit-test for this function but due to connection checks being done by `__init__()` of `MetabaseSource` I need to also mock http communication. Will continue working on this.

Comment (k-popov) [reply metadata-ingestion/src/datahub/ingestion/source/metabase.py]: @mayurinehate would you please check the PR once again? I've implemented the unit-test for the function and it passes normally.

I've also moved network initialization of `MetabaseSource` class into separate function which is called in `MetabaseSource.__init__()`. There are no changes in behavior but now init is logically split into "local" initialization and "network" initialization.

I see some tests are failing but these are not related to the change I made.

File: metadata-ingestion/src/datahub/ingestion/source/metabase.py
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/metabase.py:63]: We don't need this platform_instance_map here. Its already present in superclass `DatasetLineageProviderConfigBase`

Comment (k-popov) [reply metadata-ingestion/src/datahub/ingestion/source/metabase.py]: Thank you for noticing that. I've removed this definition and rebased changes onto fresh `master`. All related tests seem to be fine.

**Comment by k-popov** [comment]
@mayurinehate , thank you for your response. I've removed redundant attribute definition. Related tests look fine. Would you please check the MR once again?

**Review by mayurinehate — APPROVED** [review]

**Comment by k-popov** [comment]
@mayurinehate , sorry to bother you once again but I might be missing something.

After you approved the PR I still didn't have a button to merge it. I first thought that's because `master` branch has moved forward. I've rebased changes via GitHub's UI but as a result I only lost your approval. Am I supposed to have the merge button after approval? Or one of the project maintainers is accepting the PR and actually merges the changes?

**Comment by mayurinehate** [comment]
Hi @k-popov you don't need to do any action after approval. One of the project maintainers will merge the PR.

**Review by mayurinehate — APPROVED** [review]

**Review by hsheth2 — APPROVED** [review]
LGTM

I'm not a huge fan of having both `platform_instance_map` and `database_id_to_instance_map` -- in the future, does it make sense to deprecate the former in favor of the latter?

**Pull request merged**
